receiver/receivercreator/runner.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package receivercreator
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/open-telemetry/opentelemetry-collector/component"
+	"github.com/open-telemetry/opentelemetry-collector/config"
+	"github.com/open-telemetry/opentelemetry-collector/config/configmodels"
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"go.uber.org/zap"
+)
+
+// runner starts and stops receiver instances.
+type runner interface {
+	// start a receiver instance from its static config and discovered config.
+	start(receiver receiverConfig, discoveredConfig userConfigMap) (component.Receiver, error)
+	// shutdown a receiver.
+	shutdown(rcvr component.Receiver) error
+}
+
+// receiverRunner handles starting/stopping of a concrete subreceiver instance.
+type receiverRunner struct {
+	logger       *zap.Logger
+	nextConsumer consumer.MetricsConsumerOld
+	idNamespace  string
+	ctx          context.Context
+	host         component.Host
+}
+
+var _ runner = (*receiverRunner)(nil)
+
+// start a receiver instance from its static config and discovered config.
+func (run *receiverRunner) start(receiver receiverConfig, discoveredConfig userConfigMap) (component.Receiver, error) {
+	factory := run.host.GetFactory(component.KindReceiver, receiver.typeStr)
+
+	if factory == nil {
+		return nil, fmt.Errorf("unable to lookup factory for receiver %q", receiver.typeStr)
+	}
+
+	receiverFactory := factory.(component.ReceiverFactoryOld)
+
+	cfg, err := run.loadRuntimeReceiverConfig(receiverFactory, receiver, discoveredConfig)
+	if err != nil {
+		return nil, err
+	}
+	recvr, err := run.createRuntimeReceiver(receiverFactory, cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := recvr.Start(run.ctx, run.host); err != nil {
```

Comment:
`Start()` may call `Host.ReportFatalError()`. `ReportFatalError()` exits the process. This was considered acceptable since it would normally happen at startup when receivers are created.

Now this allows receivers to be created at any time long after the startup. It means if something goes wrong the receiver's `Start` function can exit the process during normal operation. That is not considered acceptable: https://github.com/open-telemetry/opentelemetry-collector/blob/master/CONTRIBUTING.md#do-not-crash-after-startup

We will need to wrap the host here and provide a different `ReportFatalError` implementation which does not exit the process but only logs the problem and perhaps emits a metric/trace.

Reply:
What's the purpose of ReportFatalError if Start() can return an error? Why not let the caller (the core in the usual case, receivercreator) in this case decide how to handle it?

Reply:
Start() is expected to return quickly. If the component needs more time to initialize it is expected to return and continue initialization asynchronously. ReportFatalError allows the component to report initialization failure in that case.

Reply:
I changed it to just log for now. However that's probably not sufficient if receivers are assuming when they call that they are going to be restarted (by virtue of the process dying and something like k8s restarting the exited container). In the receiver creator case should we shut down the receiver that reported a fatal error? How often, how long, etc. It starts opening up a can of worms. Actually this applies to even the case where Start() returns an error.

I think we may need to introduce some kind of lifecycle for receivers if we need to handle these kinds of scenarios. It's something I considered adding to the smart agent but hadn't gotten around to it. I think we need a way to have receivers fail to start and be able to restart them with exponential retry, etc. (Although I understand for some receivers this is unlikely to help as detailed in https://github.com/open-telemetry/opentelemetry-collector/blob/master/CONTRIBUTING.md#bad-input-handling). But for many receivers it's possible they encounter some error that can be resolved by resetting to a good known state. In addition having these kinds of explicit status of receivers and whether they're running, crashing, healthy, etc. would enhance the telemetry as you could tell whether a receiver was in a good running state or not.

Reply:
Created https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/178